inference-engine/src/transformations/src/transformations/smart_reshape/reshape_with_hc_output.cpp

```diff
@@ -0,0 +1,83 @@
+// Copyright (C) 2018-2020 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/smart_reshape/reshape_with_hc_output.hpp"
+
+#include <memory>
+#include <vector>
+
+#include <ngraph/ngraph.hpp>
+#include <ngraph/pattern/matcher.hpp>
+#include <ngraph/rt_info.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/opsets/opset4.hpp>
+
+void relax_hc_reshape_followed_by_matmul(std::shared_ptr<ngraph::opset4::MatMul> matmul, std::shared_ptr<ngraph::Node> reshape,
+                                         ngraph::Output<ngraph::Node> shape_source, std::shared_ptr<ngraph::opset4::Constant> reshape_pattern,
+                                         const bool & reshape_is_A_input) {
+    const auto &raw_idx = reshape_is_A_input ? (matmul->get_transpose_b() ? -1 : -2) : (matmul->get_transpose_a() ? -2 : -1);
+    const auto &idx = ngraph::normalize_axis(matmul->description(), raw_idx, reshape->get_output_partial_shape(0).rank());
+
+    const auto &shape_of = std::make_shared<ngraph::opset4::ShapeOf>(shape_source);
+    const auto &C = std::make_shared<ngraph::opset4::Gather>(shape_of,
+                                                     ngraph::opset4::Constant::create(ngraph::element::i64, {1}, {idx}),
+                                                     ngraph::opset4::Constant::create(ngraph::element::i64, {}, {0}));
+    const auto &N = ngraph::opset4::Constant::create(ngraph::element::i64, {1}, {-1});
+    const auto &pattern_vector = reshape_is_A_input ? (matmul->get_transpose_a() ? ngraph::OutputVector({C, N}) :
+                                                       ngraph::OutputVector({N, C})) : (matmul->get_transpose_b() ? ngraph::OutputVector({N, C})
+                                                                                                                  : ngraph::OutputVector({C, N}));
+    const auto new_reshape_pattern = std::make_shared<ngraph::opset4::Concat>(pattern_vector, 0);
```

Comment:
Did you forget the &? In 22 line you do the same operation but use the reference :)